File: core/controllers/private.py
from fastapi import APIRouter, HTTPException
from typing import List, Tuple

from services.search import solr
from services.story import story, source, author, scraped_url
from services.user import user, feedback
from services import redis
from shared.types import source_types, author_types, story_types, scraped_url_types, data_set_types, search_types

router = APIRouter()


@router.post('/stories')
def add_stories(stories: Tuple[story_types.Story, ...]):
  redis.worker_queue.enqueue(story.insert_stories, stories)
  return {'addStoriesJob': "Job Queued"}


@router.get('/scraped', response_model=Tuple[str, ...])
def get_scraped_urls(source_name: str):
  urls = scraped_url.get_by_source_name(source_name)
  if not urls:
    raise HTTPException(status_code=404, detail="Urls not found")
  return urls

@router.post('/scraped')
def add_scraped_urls(scrapped_urls: Tuple[scraped_url_types.ScrapedUrl, ...]):
  redis.worker_queue.enqueue(scraped_url.add_scraped_urls, scrapped_urls)
  return {'add_scraped_urls': "Job Queued"}


@router.get('/sources/{language}',
            response_model=Tuple[source_types.Source, ...])
def get_sources(language: str):
  sources = source.get_all_sources(language)
  if not sources:
    raise HTTPException(status_code=404, detail="Sources not found")
  return sources


@router.get('/sources/reset')
def reset_sources():
  source.reset_sources()
  return {'result': 'Sources Reset'}


@router.get('/authors/name', response_model=author_types.Author)
def get_author_by_name(author_name: str):
  current_author = author.get_by_name(author_name)
  if not current_author:
    raise HTTPException(status_code=404, detail="Author not found")
  return current_author


@router.get('/training/{user_id}',
            response_model=List[data_set_types.RankingData])
def get_training_feedbacks(user_id: str):
  result = feedback.get_tf_training_data(user_id)
  if not result:
    raise HTTPException(status_code=404, detail="No Training Data Found")
  return result


@router.delete('/training/{user_id}')
def delete_user_feedbacks(user_id: str):
  redis.worker_queue.enqueue(feedback.remove_feedback_of_user, user_id)
  return {'feedbackJob': "Job Queued"}


@router.get('/user/ids', response_model=Tuple[str, ...])
def get_user_ids():
  return user.get_ids()


@router.get('/user/subscriptions/{language}', response_model=Tuple[dict, ...])
def get_user_subscriptions(language: str):
  subscriptions = user.get_subscriptions(language)
  if not subscriptions:
    raise HTTPException(status_code=404, detail="No subscriptions Found")
  return subscriptions


@router.get('/news/{language}',
            response_model=Tuple[story_types.ShortStory, ...])
def get_public_stories(language: str):
  return story.get_public_stories(language)


@router.get('/update-index/{language}')
def update_ranking_index(language: str):
  return story.update_tf_index(language)


@router.get('/solr/features')
def get_solr_features(search_query: search_types.SearchQuery):
  return solr.generic_search(search_query)


@router.get('/solr/reset')
def get_solr_reset():
  redis.worker_queue.enqueue(solr.delete_all)
  redis.worker_queue.enqueue(solr.refill)
  return {'Refill Solr': "Job Queued"}

File: core/internal/controllers/private.go
package controllers

import (
	"log"
	"net/http"

	"core/internal/feedback"
	"core/internal/models"
	"core/internal/story"
	"core/internal/tasks"
	"core/internal/user"

	"github.com/gorilla/mux"
	"github.com/hibiken/asynq"
)

func PrivateRoutes(r *mux.Router, queueClient *asynq.Client) *mux.Router {

	r.HandleFunc("/stories", func(w http.ResponseWriter, r *http.Request) {
		var stories []models.Story
		ok := DecodeJSON(w, r, &stories)
		if !ok {
			return
		}
		task, err := tasks.NewStoreStoriesTask(stories)
		if err != nil {
			RespondWithError(w, 500, "Could not create NewStoreStoriesTask")
			return
		}
		info, err := queueClient.Enqueue(task)
		if err != nil {
			RespondWithError(w, 500, "Could not enqueue NewStoreStoriesTask")
			return
		}
		log.Printf("enqueued task: id=%s queue=%s", info.ID, info.Queue)
		RespondWithJSON(w, http.StatusOK, map[string]string{"AddStoriesJob": "Job Queued"})
	}).Methods("POST")

	r.HandleFunc("/scraped", func(w http.ResponseWriter, r *http.Request) {
		sourceName := r.URL.Query().Get("source_name")
		scrapedUrls, err := story.GetScrapedUrlsBySourceName(sourceName)
		if err != nil {
			RespondWithError(w, 500, err.Error())
			return
		}
		RespondWithJSON(w, http.StatusOK, scrapedUrls)
	}).Methods("GET")

	r.HandleFunc("/scraped", func(w http.ResponseWriter, r *http.Request) {
		var scrapedUrls []models.ScrapedUrl
		ok := DecodeJSON(w, r, &scrapedUrls)
		if !ok {
			return
		}
		task, err := tasks.NewAddScrapedUrlsTask(scrapedUrls)
		if err != nil {
			RespondWithError(w, 500, "Could not create NewAddScrapedUrlsTask")
			return
		}
		info, err := queueClient.Enqueue(task)
		if err != nil {
			RespondWithError(w, 500, "Could not enqueue NewAddScrapedUrlsTask")
			return
		}
		log.Printf("enqueued task: id=%s queue=%s", info.ID, info.Queue)
		RespondWithJSON(w, http.StatusOK, map[string]string{"AddScrapedUrls": "Job Queued"})
	}).Methods("POST")

	r.HandleFunc("/sources/{language}", func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		language := vars["language"]
		sources, err := story.GetAllSources(language)
		if err != nil {
			RespondWithError(w, 500, err.Error())
			return
		}
		RespondWithJSON(w, http.StatusOK, sources)
	}).Methods("GET")

	r.HandleFunc("/sources/reset", func(w http.ResponseWriter, r *http.Request) {
		ok := story.ResetSources()
		if ok {
			RespondWithJSON(w, http.StatusOK, map[string]string{"message": "Successfully reset sources"})
			return
		}
		RespondWithError(w, http.StatusBadRequest, "Failed to reset sources")
	}).Methods("GET")

	r.HandleFunc("/authors/name", func(w http.ResponseWriter, r *http.Request) {
		authorName := r.URL.Query().Get("author_name")
		author, err := story.GetAuthorByName(authorName)
		if err != nil {
			RespondWithError(w, 404, "Author not found")
			return
		}
		RespondWithJSON(w, http.StatusOK, author)
	}).Methods("GET")

	r.HandleFunc("/training/{user_id}", func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		userId := vars["user_id"]
		data, err := feedback.GetTFTrainingData(userId)
		if err != nil {
			RespondWithError(w, 500, err.Error())
			return
		}
		if len(data) == 0 {
			RespondWithError(w, 404, "No Training Data Found")
			return
		}
		RespondWithJSON(w, http.StatusOK, data)
	}).Methods("GET")

	r.HandleFunc("/training/{user_id}", func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		userId := vars["user_id"]
		task, err := tasks.NewDeleteUserFeedbackTask(userId)
		if err != nil {
			RespondWithError(w, 500, "Could not create NewDeleteUserFeedbackTask")
			return
		}
		info, err := queueClient.Enqueue(task)
		if err != nil {
			RespondWithError(w, 500, "Could not enqueue NewDeleteUserFeedbackTask")
			return
		}
		log.Printf("enqueued task: id=%s queue=%s", info.ID, info.Queue)
		RespondWithJSON(w, http.StatusOK, map[string]string{"feedbackJob": "Job Queued"})
	}).Methods("DELETE")

	r.HandleFunc("/update-index/{language}", func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		language := vars["language"]
		data, err := story.UpdateTFIndex(language)
		if err != nil {
			RespondWithError(w, 500, err.Error())
			return
		}
		RespondWithJSON(w, http.StatusOK, data)
	}).Methods("GET")

	r.HandleFunc("/user/ids", func(w http.ResponseWriter, r *http.Request) {
		ids := user.GetUserIds()
		if len(ids) == 0 {
			RespondWithError(w, 404, "No User Ids Found")
			return
		}
		RespondWithJSON(w, http.StatusOK, ids)
	}).Methods("GET")

	r.HandleFunc("/news/{language}", func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		language := vars["language"]
		subscriptions, err := story.GetPublicStories(language)
		if err != nil {
			RespondWithError(w, 500, err.Error())
			return
		}
		RespondWithJSON(w, http.StatusOK, subscriptions)
	}).Methods("GET")

	return r
}
